src/violation.rs

```diff
@@ -0,0 +1,60 @@
+use std::fmt::Debug;
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+pub trait Violation: Debug + PartialEq + Eq + Serialize + DeserializeOwned {
+    /// The message used to describe the violation.
+    fn message(&self) -> String;
+
+    /// If autofix is (potentially) available for this violation returns another
+    /// function that in turn can be used to obtain a string describing the
+    /// autofix.
+    fn autofix_formatter(&self) -> Option<fn(&Self) -> String> {
+        None
+    }
+
+    /// A placeholder instance of the violation.
+    fn placeholder() -> Self;
+}
+
+/// This trait exists just to make implementing the [`Violation`] trait more
+/// convenient for violations that can always be autofixed.
+pub trait AlwaysAutofixableViolation:
+    Debug + PartialEq + Eq + Serialize + DeserializeOwned
+{
+    /// The message used to describe the violation.
+    fn message(&self) -> String;
+
+    /// The message used to describe the available autofix.
+    fn autofix_description(&self) -> String;
```

Comment:
Nit: should this be `autofix_message` or `fix_message`, for consistency with `message` (and with the docstring)?

Reply:
Good point. I think I favor `description` over `message` because "message" begs the question "message to whom and about what?" So I think I'd rather rename `message` to `description`, to have `description` and `autofix_description`.

Reply:
Although looking at the Language Server Protocol it does have `Diagnostic.message` ... so message does probably make sense ... LSP also has `CodeAction.title` so we could name the second function `autofix_title`. WDYT?

Reply:
Changed to `autofix_title` ... I think it nicely emphasizes that the description should not be too detailed.

Reply:
Yeah that sounds reasonable. I'm happy to mirror the LSP conventions where they map nicely.